tests/Unit/Emogrifier/HtmlProcessor/AbstractHtmlProcessorTest.php

```diff
@@ -302,36 +302,138 @@ public function notAddsSecondContentTypeMetaTag()
         static::assertSame(1, $numberOfContentTypeMetaTags);
     }
 
+    /**
+     * @return string[][]
+     */
+    public function xmlSelfClosingTagDataProvider()
+    {
+        return [
+            '<br>' => ['<br/>', 'br'],
+            '<wbr>' => ['foo<wbr/>bar', 'wbr'],
+            '<embed>' => [
+                '<embed type="video/mp4" src="https://example.com/flower.mp4" width="250" height="200"/>',
+                'embed',
+            ],
+            '<picture> with <source> and <img>' => [
+                '<picture><source srcset="https://example.com/flower-800x600.jpeg" media="(min-width: 600px)"/>'
+                    . '<img src="https://example.com/flower-400x300.jpeg"/></picture>',
+                'source',
+            ],
+            '<video> with <track>' => [
+                '<video controls width="250" src="https://example.com/flower.mp4">'
+                    . '<track default kind="captions" srclang="en" src="https://example.com/flower.vtt"/></video>',
+                'track',
+            ],
+        ];
+    }
+
+    /**
+     * @return string[][]
+     */
+    public function nonXmlSelfClosingTagDataProvider()
+    {
+        return \array_map(
+            function (array $dataset) {
+                $dataset[0] = \str_replace('/>', '>', $dataset[0]);
+                return $dataset;
+            },
+            $this->xmlSelfClosingTagDataProvider()
+        );
+    }
+
+    /**
+     * @return string[][] Each dataset has three elements in the following order:
+     *         - HTML with non-XML self-closing tags (e.g. "...<br>...");
+     *         - The equivalent HTML with XML self-closing tags (e.g. "...<br/>...");
+     *         - The name of a self-closing tag contained in the HTML (e.g. "br").
+     */
+    public function selfClosingTagDataProvider()
+    {
+        return \array_map(
+            function (array $dataset) {
+                \array_unshift($dataset, \str_replace('/>', '>', $dataset[0]));
+                return $dataset;
+            },
+            $this->xmlSelfClosingTagDataProvider()
+        );
+    }
+
+    /**
+     * @return string[][]
+     */
+    public function documentTypeAndSelfClosingTagDataProvider()
+    {
+        $documentTypeDatasets = $this->documentTypeDataProvider();
+        $selfClosingTagDatasets = $this->selfClosingTagDataProvider();
+        $datasets = [];
+        foreach ($documentTypeDatasets as $documentTypeDatasetName => $documentTypeDataset) {
+            foreach ($selfClosingTagDatasets as $selfClosingTagDatasetName => $selfClosingTagDataset) {
+                $datasets[$documentTypeDatasetName . ' & ' . $selfClosingTagDatasetName]
+                    = \array_merge($documentTypeDataset, $selfClosingTagDataset);
+            }
+        }
```

Comment:
That is quite a lot of logic for unit tests. Can we simplify this?

Reply:
Simplest would be with the data copied and pasted, but that would be least maintainable and most error-prone, which is what I'm trying to avoid.

Would adding an intermediate helper method such as `joinDatasets()` make this clearer?  Then we would have
```
    /**
     * @return string[][]
     */
    public function documentTypeAndSelfClosingTagDataProvider()
    {
        return static::joinDatasets($this->documentTypeDataProvider(), $this->selfClosingTagDataProvider());
    }
```
It also occurs to me that the tests should cover each document type with at least one self-closing tag, and each self-closing tag with at least one document type, but don't necessarily need to cover all possible combinations...

Reply:
> Would adding an intermediate helper method such as joinDatasets() make this clearer? 

Yes, I think so.